### PR TITLE
Discovery with suites

### PR DESCRIPTION
**Pull request by bukzor**
This is untested as yet, but I'll test it before asking to merge.

### CODE REVIEW

File: testify/test_rerunner.py [hunk 1]
```diff
@@ -26,3 +38,35 @@ def discover(self):
...
+
+
+    def _construct_json_test(self, module, cls, methods)
+        cls = test_discovery.import_test_class(module, cls)
+        yield self._construct_test(
+            cls,
+            name_overrides=methods,
+            suites_include=self.suites_include,
+            suites_exclude=self.suites_exclude,
+            suites_require=self.suites_require,
```
Comment (bukzor) [inline testify/test_rerunner.py]: It's essential that the --require/exclude-suites options work now

File: testify/test_rerunner.py [hunk 2]
```diff
@@ -26,3 +38,35 @@ def discover(self):
...
+                prev_test = test
+                methods = [prev_test.method]
+        yield self._construct_json_test(prev_test.module, prev_test.cls, methods)
+
+    def _parse_json_test(self, test):
+        test = json.loads(test)
+        name, suites = test['name'], test['suites']
+        module, _, cls = name.partition(' ')
+        cls, _, method = cls.rpartition('.')
+        return Test(module, cls, method)
```
Comment (bukzor) [inline testify/test_rerunner.py]: I'll likely move this to Test.from_name and share the code above.

File: testify/test_rerunner.py [hunk 3]
```diff
@@ -26,3 +38,35 @@ def discover(self):
...
+        methods = [prev_test.method]
+        for test in self.rerun_test_file:
+            test = self._parse_json_test(test)
+            if test == prev_test._replace(method=test.method):
+                methods.append(test.method)
+            else:
+                yield self._construct_json_test(prev_test.module, prev_test.cls, methods)
+                prev_test = test
+                methods = [prev_test.method]
+        yield self._construct_json_test(prev_test.module, prev_test.cls, methods)
```
Comment (bukzor) [inline testify/test_rerunner.py]: These dozen lines are a mess.
I'd like to simplify while preserving the as-you-go aspect.

File: testify/test_rerunner.py [hunk 4]
```diff
@@ -16,7 +19,16 @@ def __init__(self, *args, **kwargs):
...
+        first_char = self.rerun_test_file.read(1)
+        if first_char == '{':
+            return self.discover_json()
+        else:
+            return self.discover_txt(first_char)
+
+    def discover_text(self, first_char):
+        rerun_tests = (first_char + self.rerun_test_file.read())
+        rerun_tests = rerun_tests.splitlines()
+        for class_path, lines in groupby(rerun_tests, lambda line: line.rpartition('.')[0]):
```
Comment (bukzor) [inline testify/test_rerunner.py]: This was already eagerly reading the input, so I think it's not-worse.

File: testify/test_rerunner.py [hunk 5]
```diff
@@ -16,7 +19,16 @@ def __init__(self, *args, **kwargs):
         super(TestRerunner, self).__init__(*args, **kwargs)
 
     def discover(self):
-        for class_path, lines in groupby(self.rerun_test_file, lambda line: line.rpartition('.')[0]):
+        first_char = self.rerun_test_file.read(1)
```
Comment (asottile) [inline testify/test_rerunner.py]: just read the entire file eagerly and do the splitlines here.  It simplifies your code significantly

File: testify/test_rerunner.py (hunk 3 above)
Comment (asottile) [inline testify/test_rerunner.py]: this function was complicated so I didn't really read it

File: testify/test_runner.py [hunk 6]
```diff
@@ -251,9 +251,20 @@ def get_tests_for_suite(self, selected_suite_name):
...
+    def list_tests(self, format, selected_suite_name=None):
         """Lists all tests, optionally scoped to a single suite."""
         for test in self.get_tests_for_suite(selected_suite_name):
-            print(self.get_test_method_name(test))
+            name = self.get_test_method_name(test)
+            if format == 'txt':
+                print(name)
+            elif format == 'json':
+                testcase = test.__self__
+                import json
```
Comment (asottile) [inline testify/test_runner.py]: this repository imports things at the top, follow that

File: testify/test_runner.py [hunk 7]
```diff
@@ -251,9 +251,20 @@ def get_tests_for_suite(self, selected_suite_name):
...
-    def list_tests(self, selected_suite_name=None):
+    def list_tests(self, format, selected_suite_name=None):
         """Lists all tests, optionally scoped to a single suite."""
         for test in self.get_tests_for_suite(selected_suite_name):
-            print(self.get_test_method_name(test))
+            name = self.get_test_method_name(test)
+            if format == 'txt':
+                print(name)
+            elif format == 'json':
+                testcase = test.__self__
```
Comment (asottile) [inline testify/test_runner.py]: I think this changes py2/py3

File: testify/test_rerunner.py [hunk 8]
```diff
@@ -1,11 +1,14 @@
 from __future__ import absolute_import
 from itertools import groupby
 import sys
+from collections import namedtuple
 
 from . import test_discovery
 from .test_runner import TestRunner
 
 
+Test = namedtuple('Test', 'module cls, method')
```
Comment (asottile) [inline testify/test_rerunner.py]: mixing spaces, commas and stringly typed :)  I'd prefer `namedtuple('Test', ('module', 'cls', 'method'))`

Comment (bukzor) [inline testify/test_rerunner.py]: I had thought that the string form was the passthru path, but I was wrong:
https://github.com/python/cpython/blob/2.7/Lib/collections.py#L329

File: testify/test_runner.py (hunk 7 above)
Comment (bukzor) [inline testify/test_runner.py]: Other function attributes changed, but it looks like not __self__:

```
$ python3.5
>>> class C:
...  def f(self): pass
...
>>> c = C()
>>> c
<__main__.C object at 0x7f5d099e8080>
>>> c.f.__self__
<__main__.C object at 0x7f5d099e8080>
```

File: test/test_suites_test.py [hunk 9]
```diff
@@ -36,30 +27,46 @@ class SubTestCase(SuperTestCase):
...
-            @suite('super')
-            def test_thing(self):
-                pass
+        super_instance = SuperDecoratedTestCase()
+        sub_instance = SubDecoratedTestCase()
 
-        class SubTestCase(SuperTestCase):
-            __test__ = False
+        assert_equal(super_instance.test_thing._suites, set(['super']))
+        assert_equal(sub_instance.test_thing._suites, set(['sub']))
```
Comment (asottile) [inline test/test_suites_test.py]: set literals?

File: testify/test_program.py [hunk 10]
```diff
@@ -288,10 +294,12 @@ def run(self):
         )
 
         if self.runner_action == ACTION_LIST_SUITES:
-            runner.list_suites()
+            suite_counts = runner.list_suites()
+            pp = pprint.PrettyPrinter(indent=2)
+            print(pp.pformat(dict(suite_counts)))
```
Comment (asottile) [inline testify/test_program.py]: hmmm, what is this?

File: testify/test_runner.py [hunk 11]
```diff
@@ -198,12 +198,7 @@ def list_suites(self):
             for test_method in test_instance.runnable_test_methods():
                 for suite_name in test_instance.suites(test_method):
                     suites[suite_name].append(test_method)
-        suite_counts = dict((suite_name, "%d tests" % len(suite_members)) for suite_name, suite_members in suites.items())
-
-        pp = pprint.PrettyPrinter(indent=2)
-        print(pp.pformat(dict(suite_counts)))
-
-        return suite_counts
+        return dict((suite_name, "%d tests" % len(suite_members)) for suite_name, suite_members in suites.items())
```
Comment (asottile) [inline testify/test_runner.py]: dictcomp?

File: testify/test_program.py (hunk 10 above)
Comment (bukzor) [inline testify/test_program.py]: list_suites previously had a side-effect of printing its result

File: testify/test_runner.py [hunk 12]
```diff
@@ -212,9 +207,22 @@ def get_tests_for_suite(self, selected_suite_name):
...
-            print(self.get_test_method_name(test))
+            name = self.get_test_method_name(test)
+            if format == 'txt':
+                print(name)
+            elif format == 'json':
+                testcase = test.__self__
+                print(json.dumps(
+                    dict(
+                        test=name,
+                        suites=sorted(testcase.suites(test)),
```
Comment (asottile) [inline testify/test_runner.py]: Can the method itself have suites?  The `@T.suite` decorator can apply to it, but does it do anything?  Should it raise maybe if it doesn't?

Comment (bukzor) [inline testify/test_runner.py]: Yes. methods have suites.
This is very common in yelp-main.

File: testify/test_runner.py (hunk 11 above)
Comment (bukzor) [inline testify/test_runner.py]: This is old code.
I'd rather not touch it.
py27'izing has zero value imho.

File: test/test_suites_test.py (hunk 9 above)
Comment (bukzor) [inline test/test_suites_test.py]: This is old code.
I'd rather not touch it.
py27'izing has zero value imho.

File: testify/test_runner.py (hunk 12 above)
Comment (asottile) [inline testify/test_runner.py]: ah I see it now, you're calling `testcase.suites`

File: testify/test_runner.py (hunk 11 above)
Comment (asottile) [inline testify/test_runner.py]: I mean, you're already touching it

**Comment by bukzor** [comment]
No blockers; please to merge.

**Pull request merged**